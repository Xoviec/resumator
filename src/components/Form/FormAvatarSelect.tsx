import { VoidFunctionComponent } from "react";
import { Controller, useFormContext } from "react-hook-form";
import {
  Box,
  FormControl,
  FormControlProps,
  InputLabel,
  OutlinedInput,
} from "@mui/material";
import { avatars } from "../../assets/images/avatars/avatars";
import { colors } from "../../config/theme";

interface FormAvatarSelectProps extends FormControlProps {
  name: string;
  label: string;
}

export const FormAvatarSelect: VoidFunctionComponent<FormAvatarSelectProps> = ({
  name,
  label,
  ...props
}) => {
  const { control } = useFormContext();

  return (
    <Controller
      defaultValue="1"
      control={control}
      name={name!}
      render={({field: { onChange, value }}) => (
        <FormControl fullWidth size="small" {...props}>
          <InputLabel shrink variant="outlined" htmlFor="form-avatar-selector-input">
            {label}
          </InputLabel>
          <OutlinedInput
            fullWidth
            notched
            id="form-avatar-selector-input"
            label={label}
            inputComponent={() => (
              <Box
                display="flex"
                flexDirection="row"
                flexWrap="wrap"
                justifyContent="center"
              >
                {avatars.map((avatar, i) => (
                  <Box
                    component="span"
                    sx={{
                      width: "100px",
                      height: "200px",
                      margin: "16px",
                      backgroundRepeat: "no-repeat",
                      backgroundSize: "cover",
                      cursor: "pointer",
                      backgroundImage: `url(${avatar.img})`,

                      "&:hover": {
                        outline: `2px auto ${colors.darkBlue}`,
                      },

                      ...(avatar.name === value
                        ? { outline: `2px auto ${colors.orange}` }
                        : {}),
                    }}
                    key={i}
                    onClick={() => onChange(avatar.name)}
                  />
                ))}
              </Box>
            )}
          />
        </FormControl>
      )}
    />
  );
};
