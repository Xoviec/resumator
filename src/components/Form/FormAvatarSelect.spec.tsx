import { render, fireEvent } from "@testing-library/react";
import { FunctionComponent, useEffect } from "react";
import { FormProvider, useForm } from "react-hook-form";
import { FormAvatarSelect } from "./FormAvatarSelect";

describe("FormAvatarSelect", () => {
  it("renders correctly", () => {
    const FormProviderWrapper: FunctionComponent = ({ children }) => {
      const form = useForm();

      return <FormProvider {...form}>{children}</FormProvider>;
    };

    const { asFragment } = render(
      <FormProviderWrapper>
        <FormAvatarSelect name="test-form-name" label="test-form-label" />
      </FormProviderWrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("updates the value when selected", () => {
    const fieldName = "test-form-name";

    const FormProviderWrapper: FunctionComponent<any> = ({ children, onChange }) => {
      const form = useForm();

      const fieldValue = form.watch(fieldName);

      useEffect(() => {
        onChange(fieldValue);
      }, [fieldValue, onChange]);

      return <FormProvider {...form}>{children}</FormProvider>;
    };

    const mockChange = jest.fn();

    const { getByTestId } = render(
      <FormProviderWrapper onChange={mockChange}>
        <FormAvatarSelect name={fieldName} label="test-form-label" />
      </FormProviderWrapper>
    );

    const avatar3 = getByTestId("avatar-3");
    fireEvent.click(avatar3);

    expect(mockChange).toHaveBeenCalledWith("3");
  });
});
