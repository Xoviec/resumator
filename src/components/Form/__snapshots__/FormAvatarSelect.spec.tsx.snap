// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormAvatarSelect renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for="form-avatar-selector-input"
    >
      test-form-label
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <div
        class="MuiBox-root css-1d3yqbi"
      >
        <span
          class="MuiBox-root css-1khc54h"
          data-testid="avatar-0"
        />
        <span
          class="MuiBox-root css-pu0jl8"
          data-testid="avatar-1"
        />
        <span
          class="MuiBox-root css-1f4nr86"
          data-testid="avatar-2"
        />
        <span
          class="MuiBox-root css-ciwohl"
          data-testid="avatar-3"
        />
        <span
          class="MuiBox-root css-15x9f6m"
          data-testid="avatar-4"
        />
        <span
          class="MuiBox-root css-1iz5n0q"
          data-testid="avatar-5"
        />
        <span
          class="MuiBox-root css-1uvy6sx"
          data-testid="avatar-6"
        />
        <span
          class="MuiBox-root css-10t5gb3"
          data-testid="avatar-7"
        />
        <span
          class="MuiBox-root css-xd2xro"
          data-testid="avatar-8"
        />
      </div>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1z7n62"
        >
          <span>
            test-form-label
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
